### DISCUSSION TITLE
Locale cookie may not be updated (Vercel)

### DISCUSSION THREAD

**Discussion by rodrigovcortezi** [amannn/next-intl#936] [Q&A] [answered]
When using static rendering in Vercel, further requests to the same page will get a 304 response that lacks the `set-cookie` header. Because of that, if the user enters a page with a different locale (locale prefix in pathname) the NEXT_LOCALE cookie is not updated.

This problem can be reproduced in the [App Router Example](https://next-intl-example-app-router.vercel.app/) available in the docs:
1. Open dev tools network tab.
2. Make at least two requests to `/en` (using URL) so that you start getting 304 responses.
3. Make at least two requests to `/de` (using URL) so that you start getting 304 responses for this locale too.
4. Now go to `/en` again (using URL) and verify that the NEXT_LOCALE cookie is still `de`.

I verified that the middleware is executed any way, but I believe Vercel strips the `set-cookie` header on 304.
I wonder if next-intl could avoid this issue somehow or if this is a problem that should be addressed in the hosting configuration. I couldn't change how vercel treats this so I came up with a workaround.

As a workaround for this, I created a client component with the sole responsibility of updating the cookie on first load and placed it on the layout:
```javascript
// components/syncLocale.tsx
export const SyncLocale = ({ locale, path }: SyncLocaleProps) => {
  useEffect(() => {
    document.cookie = `${COOKIE_LOCALE_NAME}=${locale}; path=${path}; max-age=${COOKIE_MAX_AGE}; sameSite=${COOKIE_SAME_SITE}`
  }, [locale, path])

  return null
}

//app/[locale]/layout.tsx
export default function RootLayout({
  children,
  params: { locale },
}: Readonly<{
  children: React.ReactNode
  params: { locale: string }
}>) {
  unstable_setRequestLocale(locale)

  return (
    <html lang={locale}>
      <SyncLocale locale={locale} path="/" />
      <body>
        // ...
      </body>
    </html>
  )
}
```

Any suggestions / experiences on this would be much appreciated!

**Comment by amannn** [comment, marked as answer]
Hey @rodrigovcortezi, thanks for the great question!

I've noticed a similar issue when the cache from the client-side router is used and have addressed this in https://github.com/amannn/next-intl/pull/790. The result is that now every time a navigation API from `next-intl` is used to switch the locale, the cookie value is kept in sync, regardless of whether the middleware is hit or not.

But back to your issue: Vercel can return a 304 response for cached routes. However, a 304 status code (that lacks the `set-cookie` response header) can only be issued if the user has visited this very route in the browser previously and the browser sends an `If-None-Match` request header along with the request.

This means that if the user starts on `/de`, without having visited the site before, then the user will always receive a status code 200 and never 304. When the user then manually navigates to `/en`, a status code 200 along with `set-cookie` is returned (since this is a different route, no `if-none-match` header is added). Only if the user then manually navigates to `/de` again, a 304 status code will be returned without `set-cookie`. If the user however uses the provided controls in the app to switch the language, the cookie value is kept in sync (due to #790).

So while ideally I'd really like to avoid this problem, I'm unsure if practically users will really run into this.

That being said, if there's a way `next-intl` could address this internally, I'd definitely be happy to! The issue is somewhat that `next-intl` consists of different parts and users can mix and match and use as little as just the middleware from the library. Therefore there's unfortunately not really a central piece (not even `i18n.ts`) that is guaranteed to run and could potentially inject client-side logic to reliably update the cookie.

Your fix with `SyncLocale` looks solid though, so if this aspect is important to you, that's surely a good idea!

Hope this provides some further context!

**Reply by rodrigovcortezi** [reply]
Hey @amannn, really appreciate the detailed answer. Also, big thanks for creating `next-intl`. Super useful!

I agree it seems like manually changing the locale doesn’t happen all that ofen. However, If the intended behavior is to save the last used locale for next user navigations, this particular use case seems to be uncovered. I think addressing this issue would increase the consistency of the package behavior.

Of course, I understand your points about the complexities and considerations involved in addressing this issue whitin `next-intl`. But just as an ideia (low confidence on this), I wonder if centralizing cookie management on the client side might offer a more streamlined and reliable approach:

1. Centralized cookie management logic.
2. Reliable cookie synchronization. (Solves both 304 and router cache problems)
3. Enable prefetch of different locales. (Disabled because of server cookie management)

As a downside, this probably would introduce an addional step for the package usage setup.

Anyways.. For the time being, I will use the client component I showed in the example.
Thanks again for the engaging discussion!

**Reply by amannn** [reply]
> I wonder if centralizing cookie management on the client side might offer a more streamlined and reliable approach

Do you mean if `next-intl` would expose a `SyncLocale` component like the one from you that needs to be added as part of the setup? Hmm, yeah I agree it could bring additional consistency, but it works only after hydration—so if you try really hard on a flaky network, you might still run into cases where the cookie becomes out of sync 😄. There's a slight advantage with the `set-cookie` header here. The 304 situation is furthermore also only the case for purely static routes (and on certain hosting platforms), I'm wondering if PPR will bring something new to the table here where dynamic rendering becomes more common and a 304 response code less frequent for these routes. I have to say, being able to prefetch different locales is a nice side effect, but here I'm wondering if changing the locale is something that needs particular optimization in regard to performance.

All in all, I think for the time being, `next-intl` is in a "good enough" situation from my perspective but I'll keep this in mind in case similar questions come up and might re-evaluate in the future!

Thanks again for your thoughts, I really appreciate it! 🙏
